### PR TITLE
Refactored DiffusionArguments for new root peers configuration

### PR DESCRIPTION
**Pull request by bolt12**

### CODE REVIEW

File: ouroboros-network/src/Ouroboros/Network/PeerSelection/RootPeersDNS.hs [hunk 1]
```diff
@@ -444,19 +447,17 @@ localRootPeersProvider tracer
             Right results -> do
               atomically $ do
                 rootPeersGroups <- readTVar rootPeersGroupsVar
-                let rootPeersGroups' = IntMap.fromList $ zip [0,1..] rootPeersGroups
-                    (target, entry)  = rootPeersGroups' IntMap.! group
+                let (target, entry)  = rootPeersGroups `Seq.index` index
                     resultsMap       = Map.fromList (map fst results)
                     entry'           = entry <> resultsMap
```
Comment (coot) [inline ouroboros-network/src/Ouroboros/Network/PeerSelection/RootPeersDNS.hs:452]: ^ shouldn't this be `resultsMap <> entry`, `Map.union` is left biased?

File: ouroboros-network/src/Ouroboros/Network/PeerSelection/RootPeersDNS.hs [hunk 2]
```diff
@@ -319,8 +320,8 @@ newResolverResource resolvConf = go
 --
 
 data TraceLocalRootPeers =
-       TraceLocalRootDomains [(Int, Map RelayAddress PeerAdvertise)]
-       -- ^ 'Int' is the configured valency for the local producer groups
+     TraceLocalRootDomains [(Int, Map RelayAddress PeerAdvertise)]
+     -- ^ 'Int' is the configured valency for the local producer groups
```
Comment (coot) [inline ouroboros-network/src/Ouroboros/Network/PeerSelection/RootPeersDNS.hs:324]: I think the original indentation was better.

File: ouroboros-network/src/Ouroboros/Network/PeerSelection/RootPeersDNS.hs (hunk 1 above)
Comment (bolt12) [reply ouroboros-network/src/Ouroboros/Network/PeerSelection/RootPeersDNS.hs:452]: Ah! I think you're right, good catch! Thank you!

File: ouroboros-network/src/Ouroboros/Network/Diffusion.hs [hunk 3]
```diff
@@ -277,10 +275,9 @@ data DiffusionArguments = DiffusionArguments {
     , daPeerSelectionTargets :: PeerSelectionTargets
       -- ^ selection targets for the peer governor
 
-    , daStaticLocalRootPeers :: [(Socket.SockAddr, PeerAdvertise)]
-    , daLocalRootPeers       :: [(DomainAddress, PeerAdvertise)]
-    , daPublicRootPeers      :: [DomainAddress]
-    , daUseLedgerAfter       :: UseLedgerAfter
+    , daLocalRootPeersVar    :: StrictTVar m [(Int, Map RelayAddress PeerAdvertise)]
+    , daPublicRootPeersVar   :: StrictTVar m [RelayAddress]
+    , daUseLedgerAfterVar    :: StrictTVar m UseLedgerAfter
```
Comment (coot) [inline ouroboros-network/src/Ouroboros/Network/Diffusion.hs:280]: Do we write to all three `TVar`s? Or all some of them read only as far as network is concerned?

Comment (coot) [reply ouroboros-network/src/Ouroboros/Network/Diffusion.hs:280]: This patch is more than updating `ouroboros-consensus`, as indicated in the git message.

File: ouroboros-network/test/Test/Ouroboros/Network/PeerSelection/Instances.hs [hunk 4]
```diff
@@ -81,6 +86,42 @@ instance Arbitrary PeerSelectionTargets where
...
+             <*> arbitrary
+
+instance Arbitrary RelayAddress where
+  arbitrary = do
+    isDomain <- arbitrary
+    if isDomain
+      then RelayDomain <$> arbitrary
+      else RelayAddress
+             <$> oneof [genIPv4, genIPv6]
+             <*> (fromIntegral <$> (arbitrary :: Gen Int))
```
Comment (coot) [inline ouroboros-network/test/Test/Ouroboros/Network/PeerSelection/Instances.hs:124]: or
```
oneof [ RelayDomain <$> arbitrary
      , RelayAddresss <$> oneof [genIPv4, genIPv6]
                      <*> (fromIntegral @Int <$> arbitrary)
      ]
```

File: ouroboros-network/src/Ouroboros/Network/Diffusion.hs (hunk 3 above)
Comment (bolt12) [reply ouroboros-network/src/Ouroboros/Network/Diffusion.hs:280]: We do write to all `TVar`s when one triggers the hot config update via SIGHUP

Comment (bolt12) [reply ouroboros-network/src/Ouroboros/Network/Diffusion.hs:280]: I will update the commit and make it more descriptive!

File: ouroboros-network/test/Test/Ouroboros/Network/PeerSelection/Instances.hs (hunk 4 above)
Comment (bolt12) [reply ouroboros-network/test/Test/Ouroboros/Network/PeerSelection/Instances.hs:124]: Thanks!

File: ouroboros-network/src/Ouroboros/Network/Diffusion.hs (hunk 3 above)
Comment (coot) [reply ouroboros-network/src/Ouroboros/Network/Diffusion.hs:280]: But isn't that done in `cardano-node`?

Comment (bolt12) [reply ouroboros-network/src/Ouroboros/Network/Diffusion.hs:280]: Yes, and the Diffusion Layer provides the TVars to be updated by `cardano-node`

Comment (coot) [reply ouroboros-network/src/Ouroboros/Network/Diffusion.hs:280]: it's the other way  around, 'DiffusionArguments` pass data from `cardano-node` to `ouroboros-network`.  The `TVar`'s must be created in `cardano-node` and in `ouroboros-network` we probably only need to read them.

Comment (bolt12) [reply ouroboros-network/src/Ouroboros/Network/Diffusion.hs:280]: Ah, I see. So what happens is, only `daLocalRootPeersVar` is written by `ouroboro-network` the others are read-only I believe, but can change if over at `cardano-node` someone triggers local config update (i.e. `cardano-node` writes to them).

File: ouroboros-network/test/Test/Ouroboros/Network/PeerSelection/Instances.hs [hunk 5]
```diff
@@ -81,6 +86,40 @@ instance Arbitrary PeerSelectionTargets where
...
+    DomainAddress . encodeUtf8
+      <$> elements domains
+      <*> (fromIntegral <$> (arbitrary :: Gen Int))
+    where
+      domains = [ "test1"
+                , "test2"
+                , "test3"
+                , "test4"
+                , "test5"
+                ]
```
Comment (coot) [inline ouroboros-network/test/Test/Ouroboros/Network/PeerSelection/Instances.hs:100]: Are we using just five to make it more likely to draw the same name twice, i.e. when one is generating a list of `DomainAddress`-es?

**Review by coot — APPROVED** [review]
LGTM

File: ouroboros-network/test/Test/Ouroboros/Network/PeerSelection/Instances.hs (hunk 5 above)
Comment (bolt12) [reply ouroboros-network/test/Test/Ouroboros/Network/PeerSelection/Instances.hs:100]: I think that is a good reason, however I have to admit that I just followed what was done over at cardano-node (at `cardano-node/test/Test/Cardano/Node/Gen.hs`), but with less fun naming! 

 https://hackage.haskell.org/package/hedgehog-corpus-0.2.0/docs/src/Hedgehog.Corpus.html#cooking 😛

Comment (coot) [reply ouroboros-network/test/Test/Ouroboros/Network/PeerSelection/Instances.hs:100]: Ha ha, that's a good one, I didn't know such a thing existed.

**Pull request merged**